ansible_collections/arista/avd/plugins/plugin_utils/schema/avddataconverter.py

```diff
@@ -1,238 +1,166 @@
-            {
-                "any": jsonschema._types.is_any,
-                "array": jsonschema._types.is_array,
-                "boolean": jsonschema._types.is_bool,
-                "integer": jsonschema._types.is_integer,
-                "object": jsonschema._types.is_object,
-                "null": jsonschema._types.is_null,
-                "None": jsonschema._types.is_null,
-                "number": jsonschema._types.is_number,
-                "string": jsonschema._types.is_string,
-                "dict": jsonschema._types.is_object,
-                "str": jsonschema._types.is_string,
-                "bool": jsonschema._types.is_bool,
-                "list": jsonschema._types.is_array,
-                "int": jsonschema._types.is_integer,
-                "float": _is_float,
-            }
-        ),
-    )
+    def __init__(self, avdschema):
+        self._avdschema = avdschema
+
+        self.converters = {
+            "items": self.convert_items,
+            "keys": self.convert_keys,
+        }
+
+    def convert_data(self, data, schema: dict = None, path: list = None) -> Generator:
+        if schema is None:
+            # Get fully resolved schema (where all $ref has been expanded recursively)
+            # Performs inplace update of the argument so we give an empty dict.
+            # By default it will resolve the full schema
+            schema = {}
+            resolve_errors = self._avdschema.resolve(schema)
+            for resolve_error in resolve_errors:
+                if isinstance(resolve_error, Exception):
+                    raise AristaAvdError(resolve_error)
+
+        if path is None:
+            path = []
+
+        for key, converter in self.converters.items():
+            if key not in schema:
+                # Ignore keys not in schema
+                continue
+
+            # Converters will do inplace update of data. Any returns will be yielded conversion messages.
+            yield from converter(schema[key], data, schema, path)
+
+    def convert_keys(self, keys: dict, data: dict, schema: dict, path: list):
+        """
+        This function performs conversion on each key with the relevant subschema
+        """
+        if not isinstance(data, dict):
+            return
+
+        # We run through all the regular keys first, to ensure that all data has been converted
+        # in case some of it is referenced in "dynamic_keys" below
+        for key, childschema in keys.items():
+            if key not in data:
+                # Skip key since there is nothing to convert if the key is not set in data
+                continue
+
+            # Perform type conversion of the data for the child key if required based on "convert_types"
+            if "convert_types" in childschema:
+                yield from self.convert_types(childschema["convert_types"], data, key, childschema, path + [key])
+
+            yield from self.convert_data(data[key], childschema, path + [key])
```

Comment:
Exact same code block as from line 87 to 97 - making let's use a helper function

Reply:
Will be done in later PR - the one with deprecation messages.